components/raftstore-v2/src/operation/query/lease.rs

```diff
@@ -0,0 +1,352 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+use std::sync::{Arc, Mutex};
+
+use crossbeam::channel::TrySendError;
+use engine_traits::{KvEngine, RaftEngine};
+use kvproto::{
+    errorpb,
+    kvrpcpb::ExtraOp as TxnExtraOp,
+    raft_cmdpb::{self, RaftCmdRequest, RaftCmdResponse},
+};
+use raft::Ready;
+use raftstore::{
+    errors::RAFTSTORE_IS_BUSY,
+    store::{
+        can_amend_read, cmd_resp,
+        fsm::{apply::notify_stale_req, Proposal},
+        metrics::RAFT_READ_INDEX_PENDING_COUNT,
+        msg::{ErrorCallback, ReadCallback},
+        propose_read_index, should_renew_lease,
+        util::{check_region_epoch, LeaseState},
+        ReadDelegate, ReadIndexContext, ReadIndexRequest, ReadProgress, RequestInspector,
+        Transport,
+    },
+    Error,
+};
+use slog::{debug, error, info, o, Logger};
+use tikv_util::{box_err, time::monotonic_raw_now, Either};
+use time::Timespec;
+use tracker::GLOBAL_TRACKERS;
+
+use crate::{
+    batch::StoreContext,
+    fsm::StoreMeta,
+    raft::Peer,
+    router::{
+        message::RaftRequest, CmdResChannel, PeerMsg, QueryResChannel, QueryResult, ReadResponse,
+    },
+    Result,
+};
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub(crate) fn propose_read_index<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        is_leader: bool,
+        ch: QueryResChannel,
+        now: Timespec,
+    ) -> bool {
+        poll_ctx.raft_metrics.propose.read_index.inc();
+
+        let request = req
+            .mut_requests()
+            .get_mut(0)
+            .filter(|req| req.has_read_index())
+            .map(|req| req.take_read_index());
+        let (id, dropped) = propose_read_index(self.raft_group_mut(), request.as_ref(), None);
+        if dropped && is_leader {
+            // The message gets dropped silently, can't be handled anymore.
+            notify_stale_req(self.term(), ch);
+            poll_ctx.raft_metrics.propose.dropped_read_index.inc();
+            return false;
+        }
+
+        let mut read = ReadIndexRequest::with_command(id, req, ch, now);
+        read.addition_request = request.map(Box::new);
+        self.pending_reads_mut().push_back(read, is_leader);
+        debug!(
+            self.logger,
+            "request to get a read index";
+            "request_id" => ?id,
+            "is_leader" => is_leader,
+        );
+
+        true
+    }
+
+    fn leader_read_index<T: Transport>(
+        &mut self,
+        poll_ctx: &mut StoreContext<EK, ER, T>,
+        mut req: RaftCmdRequest,
+        ch: QueryResChannel,
+        now: Timespec,
+    ) -> bool {
```

Comment:
Why need to return `bool`?

Reply:
read_index is bool (same as v1)----I expect we will return bool as well in on_query() once the propose is done, as in v1 it will need the return value to set the has_ready flag.